test/e2e/framework/todo/node/init/init.go

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package init registers node.AllNodesReady.
+package init
+
+import (
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/framework/todo/node"
+)
+
+func init() {
+	framework.NewFrameworkExtensions = append(framework.NewFrameworkExtensions,
+		func(f *framework.Framework) {
+			ginkgo.AfterEach(func() {
+				node.AllNodesReady(f.ClientSet, 3*time.Minute)
+			})
+		},
+	)
+}
```

Comment:
> The design intent / roadmap is AFAIK still undocumented... what is the end goal, and have you reached it here?

The goal is to have a smaller core framework with optional functionality moved into sub packages that a specific E2E suite may use, but doesn't have to import when not using it. I can add that in a README.md. This PR achieves that goal.

> One of the chief complaints about framework is that it's hard to know the order in which things are getting called. 

I suppose it didn't help either that there were various custom hooks and callbacks, with poor to no documentation... I agree that this extension mechanism adds some complexity, but I compensated for it a bit by removing some of those other custom hooks in https://github.com/kubernetes/kubernetes/pull/111998.

I can explain this all better in the top-level README that I'll add.

> This, to me, complicates things even further:
>
> -   the order in which these inits get imported matters

Which shouldn't matter. Each of them must be (and is) self-contained and the code that they execute during a test run runs in separate Ginkgo nodes, so even if one fails the others still execute.

>  -  the order in which the extracted blocks of code run looks, to my eyes, to be subtly different than before

There might be small changes, but I don't think they matter.

> - framework is effectively "incomplete" without these inits (or at the least its behavior without these is definitely something we don't have experience with)

True, downstream users who expect the same behavior as before need to make changes. But such downstream users are using k/k at their own risk, so this shouldn't hold us back. All of the three in-tree e2e suites behave as before.

I am one of the downstream users who decided that this risk of random breakages is worth it, and with that hat on I can say that this change seems useful and manageable.

> My understanding is you did this to break parts of framework.BeforeEach and framework.AfterEach into sub packages, though I'm less clear on why it was necessary.

If I hadn't done that, the core framework would have had a hard dependency on several of the sub packages. Using this example here, several parts of `test/e2e/framework/node` would have to be moved back into `test/e2e/framework` to avoid the `framework -> node` dependency that exists prior to this PR.

> Are you sure there's no other way of accomplishing what you're trying to do here?

I can't think of one, short of merging everything back into a single monolithic package (but that's not the goal).

What we need is a way for a test suite author to inject new functionality into each test, but without having to to modify every single test. Each test calls `framework.NewDefaultFramework`, so an extension mechanism for that seemed suitable.